discovery/kubernetes/ingress.go

```diff
@@ -0,0 +1,189 @@
+	informer cache.SharedInformer
+	store    cache.Store
+}
+
+// NewIngress returns a new ingress discovery.
+func NewIngress(l log.Logger, inf cache.SharedInformer) *Ingress {
+	return &Ingress{logger: l, informer: inf, store: inf.GetStore()}
+}
+
+// Run implements the TargetProvider interface.
+func (s *Ingress) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	// Send full initial set of pod targets.
+	var initial []*config.TargetGroup
+	for _, o := range s.store.List() {
+		tg := s.buildIngress(o.(*v1beta1.Ingress))
+		initial = append(initial, tg)
+	}
+	select {
+	case <-ctx.Done():
+		return
+	case ch <- initial:
+	}
+
+	// Send target groups for ingress updates.
+	send := func(tg *config.TargetGroup) {
+		select {
+		case <-ctx.Done():
+		case ch <- []*config.TargetGroup{tg}:
+		}
+	}
+	s.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(o interface{}) {
+			eventCount.WithLabelValues("ingress", "add").Inc()
+
+			ingress, err := convertToIngress(o)
+			if err != nil {
+				s.logger.With("err", err).Errorln("converting to Ingress object failed")
+				return
+			}
+			send(s.buildIngress(ingress))
+		},
+		DeleteFunc: func(o interface{}) {
+			eventCount.WithLabelValues("ingress", "delete").Inc()
+
+			ingress, err := convertToIngress(o)
+			if err != nil {
+				s.logger.With("err", err).Errorln("converting to Ingress object failed")
+				return
+			}
+			send(&config.TargetGroup{Source: ingressSource(ingress)})
+		},
+		UpdateFunc: func(_, o interface{}) {
+			eventCount.WithLabelValues("ingress", "update").Inc()
+
+			ingress, err := convertToIngress(o)
+			if err != nil {
+				s.logger.With("err", err).Errorln("converting to Ingress object failed")
+				return
+			}
+			send(s.buildIngress(ingress))
+		},
+	})
+
+	// Block until the target provider is explicitly canceled.
+	<-ctx.Done()
+}
+
+func convertToIngress(o interface{}) (*v1beta1.Ingress, error) {
+	ingress, isIngress := o.(*v1beta1.Ingress)
+	if !isIngress {
+		return ingress, nil
+	}
+
+	deletedState, ok := o.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("Received unexpected object: %v", o)
+	}
+	ingress, ok = deletedState.Obj.(*v1beta1.Ingress)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown contained non-Ingress object: %v", deletedState.Obj)
+	}
+	return ingress, nil
+}
+
+func ingressSource(s *v1beta1.Ingress) string {
+	return "ingress/" + s.Namespace + "/" + s.Name
+}
+
+const (
+	ingressNameLabel        = metaLabelPrefix + "ingress_name"
+	ingressLabelPrefix      = metaLabelPrefix + "ingress_label_"
+	ingressAnnotationPrefix = metaLabelPrefix + "ingress_annotation_"
+	ingressSchemeLabel      = metaLabelPrefix + "ingress_scheme"
+	ingressHostLabel        = metaLabelPrefix + "ingress_host"
+	ingressPathLabel        = metaLabelPrefix + "ingress_path"
+)
+
+func ingressLabels(ingress *v1beta1.Ingress) model.LabelSet {
+	ls := make(model.LabelSet, len(ingress.Labels)+len(ingress.Annotations)+2)
```

Comment:
Why +2?

Reply:
Agree, that should have a comment.. But I don't know, I've just copied that over from services.go. Maybe to allocate capacity already for the address label? I can also just get rid of the +2.
@fabxc Maybe you know why it's in here?

Reply:
Well, it should be +1, as you only set one additional label besides the labels+annotations. I guess you copied it from a discovery which sets two additional labels.

Reply:
No, looks to me like all roles only set the namespace as additional labels. Thought maybe something outside discovery attaches another label. Should I convert it to +1 everywhere?

Reply:
One could check the original commit, probably a second label was set there? Anyway, changing to +1 everywhere makes sense to me.

Reply:
https://groups.google.com/forum/#!forum/prometheus-users